Feature/poetry

Zutatensuppe opened the pull request:
This adds info about how to launch the tracker using [poetry](https://python-poetry.org/).
And also adds the required `pyproject.toml` and `poetry.lock`.

Using poetry the user doesn't have to clutter their system wide python installation with dependencies and also doesn't have to manually activate / create virtual envs.

It also makes sure that certain versions of the dependencies and python are used, so it will be less likely that something suddenly stops working if the installation is done in eg. a year from now.

emilianavt commented:
Hi!

I hadn't heard about poetry before, but I don't mind adding support for it. Before merging this, it might be a good idea to look at the performance of different onnxruntime versions. In the past, I sometimes found that newer versions performed worse. Of course it can also depend on specific system configurations/platforms as well.

Zutatensuppe commented:
> Hi!
> 
> I hadn't heard about poetry before, but I don't mind adding support for it. Before merging this, it might be a good idea to look at the performance of different onnxruntime versions. In the past, I sometimes found that newer versions performed worse. Of course it can also depend on specific system configurations/platforms as well.

Sounds great, thanks! 

I tried to make some benchmarks with the different onnxruntime versions, I can only check this on my linux machine now though.
The first command i ran for 30 seconds (added a break after 30 seconds of runtime) and added the output that facetracker puts out at the end (average numbers).
Additionally i just ran the facetracker with `--benchmark 1` and added the output of that as well ^^
It seems that onnxruntime 1.4.0 and 1.9.0 are the fastest? But this probably needs way more runs to be some valid statistic, I ran this only one time to get a general feeling ^^"


# onnxruntime 1.2.0

	python facetracker.py -c 0 -W 1280 -H 720 --discard-after 0 --scan-every 0 --no-3d-adapt 1 --max-feature-updates 900 -F 12 --silent 0

	Average tracking time per detected face: 23.17 ms
	Tracking time: 8.225 s
	Frames: 355

	python facetracker.py --benchmark 1

	40.87639752317269
	54.68081615010729
	68.15683863612198
	80.59203836544842
	245.87514496221252
	109.66067673318918
	141.74403472791332

# onnxruntime 1.3.0

	python facetracker.py -c 0 -W 1280 -H 720 --discard-after 0 --scan-every 0 --no-3d-adapt 1 --max-feature-updates 900 -F 12 --silent 0

	Average tracking time per detected face: 22.94 ms
	Tracking time: 8.144 s
	Frames: 355

	python facetracker.py --benchmark 1

	46.220875212238745
	54.185312834514434
	70.87965562104398
	83.20722684356413
	246.35995037068255
	107.36604017555649
	137.14637318935183

# onnxruntime 1.4.0

	python facetracker.py -c 0 -W 1280 -H 720 --discard-after 0 --scan-every 0 --no-3d-adapt 1 --max-feature-updates 900 -F 12 --silent 0

	Average tracking time per detected face: 22.46 ms
	Tracking time: 7.975 s
	Frames: 355

	python facetracker.py --benchmark 1

	39.42163772381958
	53.67244569702396
	71.88058541087034
	85.19716173303586
	249.60590434889787
	99.42712300342856
	142.5621757093456

# onnxruntime 1.6.0

	python facetracker.py -c 0 -W 1280 -H 720 --discard-after 0 --scan-every 0 --no-3d-adapt 1 --max-feature-updates 900 -F 12 --silent 0

	Average tracking time per detected face: 23.82 ms
	Tracking time: 8.433 s
	Frames: 354

	python facetracker.py --benchmark 1

	42.978261209706695
	51.96419512800446
	68.0718038281423
	82.07084708666885
	236.3178417488285
	102.43782420793403
	134.69000461126265

# onnxruntime 1.7.0

	python facetracker.py -c 0 -W 1280 -H 720 --discard-after 0 --scan-every 0 --no-3d-adapt 1 --max-feature-updates 900 -F 12 --silent 0

	Average tracking time per detected face: 23.76 ms
	Tracking time: 8.411 s
	Frames: 354

	python facetracker.py --benchmark 1

	41.36145009260007
	38.02012343220833
	67.51415081439364
	61.37397745045868
	237.98157900338225
	85.73473083004116
	135.59409920925938

# onnxruntime 1.8.0

	python facetracker.py -c 0 -W 1280 -H 720 --discard-after 0 --scan-every 0 --no-3d-adapt 1 --max-feature-updates 900 -F 12 --silent 0

	Average tracking time per detected face: 23.75 ms
	Tracking time: 8.407 s
	Frames: 354

	python facetracker.py --benchmark 1

	44.99115603682891
	52.35199659822497
	68.79999551731377
	81.72462144189943
	241.0719660972485
	94.24079452269515
	135.02491241178473

# onnxruntime 1.9.0

	python facetracker.py -c 0 -W 1280 -H 720 --discard-after 0 --scan-every 0 --no-3d-adapt 1 --max-feature-updates 900 -F 12 --silent 0

	Average tracking time per detected face: 22.56 ms
	Tracking time: 7.987 s
	Frames: 354

	python facetracker.py --benchmark 1

	47.277882090161754
	53.96294537540634
	61.57100479541408
	85.95325140707136
	241.44259409076608
	105.42615468117275
	140.82447340369237

emilianavt commented:
Thank you, that looks promising for the new versions! I'll also try and give these a run later.

emilianavt commented:
I also did some runs with the benchmark function modified to run for 2000 iterations instead of 100, but overall the variance seems to be too high to allow any real conclusion, so I think 1.9.0 should be fine.
